### DISCUSSION TITLE
API for FPU registers raw data conversion

### DISCUSSION THREAD

**Discussion by vlasta-labsky** [unicorn-engine/unicorn#1950] [Ideas]
I'm making some code that reads and writes FPU registers on X86.
Now I have this
`uint8_t st0[10]; // 80bit raw float`
`uc_reg_read(m_uc, UC_X86_REG_ST0, &st0);`
`float val = *reinterpret_cast<long double *>(st0);`
I am not 100% sure that this is a correct way to perform the conversion for all possible host platforms.
It would be nice to include API in unicorn to read/write FPU registers in platform-independent way,
to host formats like float or double.

**Comment by wtdcode** [comment]
Unfortunately, FPU registers by nature is not "platform-independent".

I will suggest you looking at `uc_reg_reg2`, which does necessary boudry checks.

**Reply by vlasta-labsky** [reply]
I am not too concerned about bounds, as for one specific guest (in this case X86_32),
the FPU internal format is going to be specific size (in this case 10 bytes).

What makes me worried is the `reinterpret_cast` to `long double`. 
From what I researched it looks like `long double` is implemented differently
on different host architectures  (fp128 on arm64) and even non-implemented on windows.
I am relying on specific gcc supplied algorithm for this conversion.
This works well for amd64 host and x86 guest, but can behave badly on other hosts.

The idea was that guest format can be known to unicorn,
and there are standard portable iee754 formats like fp32 or fp64 (that should be implemented on every host).
If unicorn would provide such algorithms in it's codebase and automaticly pick the right one,
it would be very nice.
